clients/src/main/java/org/apache/kafka/common/security/scram/ScramSaslClient.java

```diff
@@ -0,0 +1,236 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.common.security.scram;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Map;
+
+import javax.security.auth.callback.Callback;
+import javax.security.auth.callback.CallbackHandler;
+import javax.security.auth.callback.NameCallback;
+import javax.security.auth.callback.PasswordCallback;
+import javax.security.auth.callback.UnsupportedCallbackException;
+import javax.security.sasl.SaslClient;
+import javax.security.sasl.SaslClientFactory;
+import javax.security.sasl.SaslException;
+
+import org.apache.kafka.common.errors.IllegalSaslStateException;
+import org.apache.kafka.common.security.scram.ScramMessages.ClientFinalMessage;
+import org.apache.kafka.common.security.scram.ScramMessages.ServerFinalMessage;
+import org.apache.kafka.common.security.scram.ScramMessages.ServerFirstMessage;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * SaslClient implementation for SASL/SCRAM.
+ * <p>
+ * This implementation expects a login module that populates username as
+ * the Subject's public credential and password as the private credential.
+ *
+ * @see <a href="https://tools.ietf.org/html/rfc5802">RFC 5802</a>
+ *
+ */
+public class ScramSaslClient implements SaslClient {
+
+    private static final Logger log = LoggerFactory.getLogger(ScramSaslClient.class);
+
+    enum State {
+        SEND_CLIENT_FIRST_MESSAGE,
+        RECEIVE_SERVER_FIRST_MESSAGE,
+        RECEIVE_SERVER_FINAL_MESSAGE,
+        COMPLETE,
+        FAILED
+    };
+
+    private final ScramMechanism mechanism;
+    private final CallbackHandler callbackHandler;
+    private final ScramFormatter formatter;
+    private String clientNonce;
+    private State state;
+    private byte[] saltedPassword;
+    private ScramMessages.ClientFirstMessage clientFirstMessage;
+    private ScramMessages.ServerFirstMessage serverFirstMessage;
+    private ScramMessages.ClientFinalMessage clientFinalMessage;
```

Comment:
It is needed to validate server final message which arrives after the response is sent, hence the value is stored.